Together.ai demo app (#178)
dbos-inc/dbos-demo-apps: python/empyrical_evidences/README.md

<h1 align="center">Empyrical Evidences</h1>

<p align="center">
  Search Hacker News for comments relevant to your favorite academic papers.
  Powered by Together.ai and DBOS.
</p>

## What It Does

This application searches Hacker News for comments relevant to your favorite academic papers.
Specifically, it uses the Together.ai inference APIs to identify key topics in papers, then searches Hacker News for related comments and ranks them.
It is serverlessly hosted on DBOS Cloud.

## How It Works

- First, you upload an academic paper. The app uses the Together.ai API to query the paper's embeddings and store them in Postgres using pgvector.
- Then, you search Hacker News for comments relevant to the paper.
  The app does this with a multi-agent workflow that first extracts the paper's key topics, then searches Hacker News for relevant comments, then ranks the comments.

## Why Use DBOS

### Serverless

- Go from local to production cloud deployment in 30 seconds with a single command.
- [15x cheaper](https://www.dbos.dev/blog/dbos-vs-lambda-cost), [25x faster](https://www.dbos.dev/blog/dbos-vs-aws-step-functions-benchmark). 0 devops.

### Durable, transactional execution

> - Workflows are always resumed where they left off. Save money and time by not querying APIs again for already-executed steps.

Reviewer: ```suggestion
- If your app is interrupted, workflows automatically resume from where they left off. Save money and time by not querying APIs again for already-executed steps.
```

- Workflows steps have exactly-once or at-least-once guarantees.

### Observability

> - DBOS Transact applications automatically generate and export OpenTelemetry traces.

Reviewer: DBOS automatically generates and exports OpenTelemetry traces for your application

- With DBOS Cloud, access a full observability pipeline exporting and collecting your observability data automatically. Access a dashboard to query your data.

## Setup Instructions

> ### Creating A Together.ai Account

Reviewer: ```suggestion
### Creating a Together.ai Account
```

<h1 align="center">Empyrical Evidences</h1>

<p align="center">

>   Search hackernews for comment relevant to academic papers.

Reviewer: ```suggestion
  Search Hacker News for comments relevant to academic papers.
```

  Powered by Together.ai and DBOS.
</p>

> ## What does the app do

Reviewer: ```suggestion
## What It Does
```

> This application uses together.ai inference API to identity relevant topics in academic papers and rank the most relevant related comments from hackernews. It is orchestrated with DBOS Transact and demonstrates how DBOS:

Reviewer: What It Does

How It Works

- You upload papers and it embeddings
- You search with multi-agent workflow

What DBOS Does

Instructions + How To Use It

Likewise, DBOS will ensure that workflows can be retried where they left off. If the program crashes after the first two steps completed, it will only run the third step.

---

## Setting up Together.ai

### Creating a together.ai account

Go to https://api.together.ai/ and sign-up

### API key

An API key will be provided for you when you register. You can always retrieve it in your [settings page](https://api.together.ai/settings/api-keys)

## Deploying to DBOS Cloud

### Creating a DBOS Cloud account

Sign-up for [DBOS Cloud](https://console.dbos.dev/).

### Setup DBOS Cloud CLI

To deploy this app to DBOS Cloud, first install the DBOS Cloud CLI (requires Node):

```shell
npm i -g @dbos-inc/dbos-cloud
```

Then, run this command to deploy your app:

```shell
dbos-cloud app deploy
```

This command outputs a URL -- this is where your application is hosted on DBOS Cloud.
You can visit the [DBOS Cloud Console](https://console.dbos.dev/) to see your app's status and logs.

## Running Locally

To run this app locally, you need a Postgres database.
If you have Docker, you can start one with:

```shell
export PGPASSWORD=dbos
python3 start_postgres_docker.py
```

Then run:

```shell

> python3 -m venv dbos-empyrical-venv

Reviewer: Just `.venv`

<h1 align="center">Empyrical Evidences</h1>

<p align="center">
  Search Hacker News for comments relevant to your favorite academic papers.
  Powered by Together.ai and DBOS.
</p>

## What It Does

> This application uses Together.ai inference APIs to identify relevant topics in academic papers, search Hacker News for related comments and rank them.

Reviewer: ```suggestion
This application uses the Together.ai inference APIs to identify key topics in academic papers, then search Hacker News for related comments and rank them.
```

It is serverlessly hosted on DBOS Cloud.

## How It Works

- First, you upload an academic paper. The app uses the Together.ai API to query the paper's embeddings and store them in Postgres using pgvector.
- Search Hacker News. The app uses a multi-agent workflow and Together.AI inference API. First, it extracts the paper's key topics. Second, it searchs Hacker News for relevant comments. Third, it ranks the comments using Together.ai API.

## Why Use DBOS

> - Durable execution: if the application crashes, workflows will be resumed where they left off and Together.ai APIs will not be queried again for already-executed steps.

Reviewer: We'll workshop this list when you get to the office.

<h1 align="center">Empyrical Evidences</h1>

<p align="center">
  Search Hacker News for comments relevant to your favorite academic papers.
  Powered by Together.ai and DBOS.
</p>

## What It Does

This application uses Together.ai inference APIs to identify relevant topics in academic papers, search Hacker News for related comments and rank them.

> The app is hosted on DBOS Cloud.

Reviewer: ```suggestion
It is serverlessly hosted on DBOS Cloud.
```

## How It Works

> - Upload a paper. The app uses Together.ai API to query the paper's embeddings and store them in postgres/pgvector.

Reviewer: ```suggestion
- First, you upload an academic paper. The app uses the Together.ai API to query the paper's embeddings and store them in Postgres using pgvector.
```

> - Search Hacker News. The app uses a multi-agent workflow and Together.AI inference API. First, it extracts the paper's key topics. Second, it searchs Hacker News for relevant comments. Third, it ranks the comments using Together.ai API.

Reviewer: ```suggestion
- Then, you search Hacker News for comments relevant to the paper. 
The app does this with a multi-agent workflow that first extracts the paper's key topics, then searches Hacker News for relevant comments, then ranks the comments.
```

## Why Use DBOS

> - Fast Serverless deploy to the cloud.

Reviewer: Flesh this out a bit

Author: let me know how this reads now

- Durable execution: if the application crashes, workflows will be resumed where they left off and Together.ai APIs will not be queried again for already-executed steps.

> - Automatically capture provenance/lineage data for your workflows.

Reviewer: Rewrite this to say something about automatic tracing

## Setup Instructions

### Creating A Together.ai Account

Go to https://api.together.ai/ and sign-up.

### API Key

An API key will be provided for you when you register. You can always retrieve it in your [settings page](https://api.together.ai/settings/api-keys).

## Deploying To DBOS Cloud

> ### Creating A DBOS Cloud Account

Reviewer: Delete this, deploy auto-registers you now
